creusot/src/translation/specification.rs

```diff
@@ -330,22 +341,20 @@ impl<'a, 'tcx> thir::visit::Visitor<'a, 'tcx> for PurityVisitor<'a, 'tcx> {
         match expr.kind {
             ExprKind::Call { fun, .. } => {
                 if let &ty::FnDef(func_did, _) = self.thir[fun].ty.kind() {
-                    if !util::is_predicate(self.tcx, func_did)
-                        && !util::is_logic(self.tcx, func_did)
-                        && !util::get_builtin(self.tcx, func_did).is_some()
-                        && !pearlite_stub(self.tcx, self.thir[fun].ty).is_some()
-                        && !is_overloaded_item(self.tcx, func_did)
-                    {
+                    if self.in_pure_ctx != self.is_pure(fun, func_did) {
+                        let msg = if self.in_pure_ctx {
+                            "called impure program function in logical context"
+                        } else {
+                            "called logical function in impure context"
+                        };
+
                         self.tcx.sess.span_err_with_code(
```

Comment:
I think we could change `PurityVisitor` to take a `ctx` instead and use `ctx.error` to build the error.